File: datasets/C++/Mistral/Algorithm/hard2.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int chiefHopper(vector<int>& arr) {
    int n = arr.size();
    int minEnergy = 0;

    for (int i = n - 1; i >= 0; i--) {
        int currentEnergy = (minEnergy + arr[i] + 1) / 2;
        minEnergy = currentEnergy;
    }

    return minEnergy;
}

int main() {
    int n;
    cout << "Enter the number of buildings: ";
    cin >> n;

    vector<int> arr(n);
    cout << "Enter the heights of the buildings: ";
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }

    cout << "Minimum starting energy: " << chiefHopper(arr) << endl;

    return 0;
}

File: datasets/Python/Mistral/Algorithm/hard2.py
def chief_hopper(arr):
    n = len(arr)
    min_energy = 0

    for i in range(n - 1, -1, -1):
        current_energy = (min_energy + arr[i] + 1) // 2
        min_energy = current_energy

    return min_energy

# Input
n = int(input("Enter the number of buildings: "))
arr = list(map(int, input("Enter the heights of the buildings: ").split()))

# Output
print("Minimum starting energy:", chief_hopper(arr))
